crates/walrus-service/src/node/pending_sliver_cache.rs

```diff
@@ -0,0 +1,532 @@
+                    )
+                }
+            }
+            Entry::Vacant(entry) => {
+                let mut blob_entry = PendingBlobEntry::new(now);
+                Self::add_sliver(
+                    &mut blob_entry,
+                    key,
+                    sliver,
+                    new_len,
+                    &mut self.sliver_count,
+                    &mut self.total_bytes,
+                )?;
+                entry.insert(blob_entry);
+                Ok(true)
+            }
+        }
+    }
+
+    fn add_sliver(
+        entry: &mut PendingBlobEntry,
+        key: SliverCacheKey,
+        sliver: Sliver,
+        new_len: usize,
+        global_sliver_count: &mut usize,
+        global_total_bytes: &mut usize,
+    ) -> Result<bool, PendingSliverCacheError> {
+        *global_sliver_count = global_sliver_count
+            .checked_add(1)
+            .ok_or(PendingSliverCacheError::Saturated)?;
+        *global_total_bytes = global_total_bytes
+            .checked_add(new_len)
+            .ok_or(PendingSliverCacheError::Saturated)?;
+        entry.total_bytes = entry
+            .total_bytes
+            .checked_add(new_len)
+            .ok_or(PendingSliverCacheError::Saturated)?;
+        entry.slivers.insert(key, sliver);
+        Ok(true)
+    }
+
+    fn contains(&mut self, blob_id: &BlobId, key: SliverCacheKey) -> bool {
+        self.evict_expired(Instant::now());
+        self.blobs
+            .get(blob_id)
+            .is_some_and(|entry| entry.slivers.contains_key(&key))
+    }
+
+    fn has_blob(&mut self, blob_id: &BlobId) -> bool {
+        self.evict_expired(Instant::now());
+        self.blobs.contains_key(blob_id)
+    }
+
+    fn drain(&mut self, blob_id: &BlobId) -> Vec<CachedSliver> {
+        self.evict_expired(Instant::now());
+        self.blobs
+            .remove(blob_id)
+            .map(|entry| {
+                self.sliver_count = self.sliver_count.saturating_sub(entry.len());
+                self.total_bytes = self.total_bytes.saturating_sub(entry.bytes());
+                entry
+                    .slivers
+                    .into_iter()
+                    .map(|(key, sliver)| CachedSliver {
+                        sliver_pair_index: key.sliver_pair_index,
+                        sliver,
+                    })
+                    .collect()
+            })
+            .unwrap_or_default()
+    }
+
+    fn insert_many(
+        &mut self,
+        blob_id: BlobId,
+        slivers: Vec<CachedSliver>,
+    ) -> Result<(), PendingSliverCacheError> {
+        for cached in slivers {
+            let key = SliverCacheKey::new(cached.sliver_pair_index, cached.sliver.r#type());
+            self.insert(blob_id, key, cached.sliver)?;
+        }
+        Ok(())
+    }
+}
+
+/// Slivers waiting for a blob registration to be observed.
+#[derive(Debug)]
+pub(crate) struct PendingSliverCache {
+    max_sliver_bytes: usize,
+    inner: Mutex<PendingSliverCacheInner>,
```

Comment:
remark: similarly, I feel this global lock could be problematic.

What about using MokaCache<BlobId, Arc<Mutex< PendingBlobEntry >>>. I think in this case, PendingBlobEntry is mutable, and also have parallelism across blobs?